src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonDictionaryDefaultConverter.cs

```diff
@@ -0,0 +1,311 @@
+            }
+
+            return converter!;
+        }
+
+        internal override sealed bool OnTryRead(
+            ref Utf8JsonReader reader,
+            Type typeToConvert,
+            JsonSerializerOptions options,
+            ref ReadStack state,
+            out TCollection value)
+        {
+            bool shouldReadPreservedReferences = options.ReferenceHandling.ShouldReadPreservedReferences();
+
+            if (!state.SupportContinuation && !shouldReadPreservedReferences)
+            {
+                // Fast path that avoids maintaining state variables and dealing with preserved references.
+
+                if (reader.TokenType != JsonTokenType.StartObject)
+                {
+                    ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(TypeToConvert);
+                }
+
+                CreateCollection(ref state);
+
+                JsonConverter<TValue> elementConverter = GetElementConverter(ref state);
+                if (elementConverter.CanUseDirectReadOrWrite)
+                {
+                    while (true)
+                    {
+                        // Read the key name.
+                        reader.Read();
+
+                        if (reader.TokenType == JsonTokenType.EndObject)
+                        {
+                            break;
+                        }
+
+                        if (reader.TokenType != JsonTokenType.PropertyName)
+                        {
+                            ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(TypeToConvert);
+                        }
+
+                        state.Current.JsonPropertyNameAsString = reader.GetString();
+
+                        // Read the value and add.
+                        reader.Read();
+                        TValue element = elementConverter.Read(ref reader, typeof(TValue), options);
+                        Add(element, options, ref state);
+                    }
+                }
+                else
+                {
+                    while (true)
+                    {
+                        // Read the key name.
+                        reader.Read();
+
+                        if (reader.TokenType == JsonTokenType.EndObject)
+                        {
+                            break;
+                        }
+
+                        if (reader.TokenType != JsonTokenType.PropertyName)
+                        {
+                            ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(TypeToConvert);
+                        }
+
+                        state.Current.JsonPropertyNameAsString = reader.GetString();
+
+                        // Read the value and add.
+                        reader.Read();
+                        elementConverter.TryRead(ref reader, typeof(TValue), options, ref state, out TValue element);
+                        Add(element, options, ref state);
```

Comment:
I am pretty sure you are missing a call to `state.Current.EndElement` after the `Add` calls here and in line 113, but doing that breaks tests related to json path... but I doubt of the correctness of those tests.

Reply:
Please create an issue and\or fix any JSON path issues related to dictionaries.

Reply:
https://github.com/dotnet/runtime/issues/31955